File: 2019/day25/cpp/src/cryo.cpp
#include "cryo.hpp"

#include <iostream>

Cryo::Cryo(std::string file_name) : m_intcode(file_name, {}, false, false) {}

void Cryo::play() {
  std::vector<std::string> message;
  std::string row("");

  while (true) {
    auto output = m_intcode();

    if (std::get<1>(output) == -WINT_MAX) {
      // Print out the new screen
      for (auto &row_ : message) {
        std::cout << row_ << std::endl;
      }

      // Read input command and pass it to the intcode
      std::string input;
      std::getline(std::cin, input);

      for (auto &chr : input) {
        m_intcode.m_input.push_back(int(chr));
      }
      m_intcode.m_input.push_back(int('\n'));

      // Check if we want to quit
      if (char(m_intcode.m_input[0]) == 'q') {
        return;
      }

      // Clear the message
      message.clear();
      row = "";

      m_intcode.m_idle = false;

    } else {
      if (std::get<1>(output) == 10) {
        message.push_back(row);
        row = "";
      } else {
        row += char(std::get<1>(output));
      }
    }
  }
}

File: 2019/day25/python/cryo.py
from os.path import dirname
from os.path import realpath
from os.path import join
import sys

sys.path.append(join(dirname(realpath(__file__)), *["..", ".."]))
from day5.python.intcode import Intcode


class Cryo:
    """Class playing the game of navigating through the rooms"""

    def __init__(self, file):
        """Initialize the intcode"""
        self.ic = Intcode(file=file, reset=False, verbose=False)

    def play(self):
        """Use the input from the terminal to play the game"""
        message, row = [], ""
        while True:
            cond, output = self.ic()

            # Run until we are promted an input
            if output is None:

                # Print current message
                for row in message:
                    print(row)

                # Read user input
                self.ic.input = [ord(i) for i in str(input()) + "\n"]

                if (chr(self.ic.input[0]) == "q") or cond:
                    return

                # Reset message
                message, row = [], ""

                # Wake the intcode computer
                self.ic.idle = False

            # Otherwise build the screen
            else:
                # New line
                if output == 10:
                    message.append(row)
                    row = ""
                # Add to the row
                else:
                    row += chr(output)
